### ISSUE TITLE
Nearest Neighbour in the paper

### ISSUE THREAD

**Issue by HasnainRaz** [lzzcd001/OSLSM#11]
Can you guys please share how you did nearest neighbour on the fc7 features to obtain the results for the 1-NN. I'm trying to reproduce the results. But the methodology for sampling examples, computing distances and obtaining a segmentation map is not discussed in the paper.

**Comment by haamoon** [comment]
Hi HasnainRaz,

Lets say fc7 features for each image is a HxWxD tensor. Each D-dimensinoal feature is feature of a 32x32 image patch and there are HxW of them in total. The label of each patch of image in the supportset is set to +1 if most of the pixels in that patch are labeled as positive and 0 otherwise. We find all the features from both query and the supportset image patches and use l2 distance measure to find the nn.

Best,
Amirreza

**Comment by HasnainRaz** [comment]
Thank you. I was downsizing the support mask to HxW to find the labels for each element in the feature map, which is probably why I wasn't getting the same IoU numbers. I'll try counting the pixels in 32x32 patches instead, because heavy downsampling like that sometimes kills the foreground labels completely.

**Comment by HasnainRaz** [comment]
Also, do you guys count the background class in your IoU evaluation? (mean IoU over classes includes the background class?)

**Comment by haamoon** [comment]
Actually, something like bilinear downsampling should be OK in my opinion. About the other question see if this helps: https://github.com/lzzcd001/OSLSM/issues/12

**Comment by HasnainRaz** [comment]
Okay so here I summarize the 1-NN approach I am using, please feel free to point out any differences, mainly, I can't get close to the same IoU score on the splits, mine are:
For 1-shot:

| Method | Split 1  | Split 2 | Split 3 | Split 4 |
| --------|--------|------- | -------|------ |
| Yours 1-NN | 25.3 | 44.9| 41.7 | 18.5 |
| Mine 1-NN |23.3 | 32.3 | 31 | 18.4  |

For 5-shot:

| Method | Split 1  | Split 2 | Split 3 | Split 4 |
| --------|--------|------- | -------|------ |
| Yours 1-NN | 34.5 | 53 | 46.9 | 25.6 |
| Mine 1-NN | 36.7 | 43 | 46.1 | 31.4  |

The process I'm following is summarised here:

1. Train FCN on images from PASCAL and SBD annotation, remove all test class annotation from the train set annotations (set to background) for the current split. Train till convergence on 15 classes (16 including background).
2. For images in validation set, sample a test class, sample N (N=1 or 5) images for support, and 1 image for query containing that class, pass support images through FCN until FC7 layer to obtain HxWxD fc7 feature volume.
3. Downsize binarized (background=0, test class=1) support ground truth to HxW, map 1 and 0 labels to each D dimensional feature vector in the volume.
4. Pass query image through FCN, for each D dimensional feature in the query fc7 output volume, check distance with each D dimensional support feature using l2 distance. Copy appropriate label.
5. We now have a HxW segmentation map of 1s and 0s, upsample (using bilinear interpolation and thresholding) to the correct dimension as query image, evaluate IoU with the ground truth.

Is this the same as yours? Can you point out any differences?

**Comment by haamoon** [comment]
Yes, it seems similar. I realize that you are training your own model for this task. Have you tried using the model in this repository? It could be useful to find out whether the difference is in the models or the testing process? 

Also, in #1 we remove an image from the training set if it does not contain any of the training objects. I believe this might increase the generalization performance of the model for the new objects.
